advanced/Templates/pihole-FTL.service

```
@@ -9,8 +9,39 @@
 # Description:       Enable service provided by pihole-FTL daemon
 ### END INIT INFO
 
+# Get PID of main pihole-FTL process
+FTLCONFFILE="/etc/pihole/pihole-FTL.conf"
+DEFAULT_PID_FILE="/run/pihole-FTL.pid"
+getFTLPID() {
+  if [ -s "$FTLCONFFILE" ]; then
+    # if PIDFILE is not set in pihole-FTL.conf, use the default path
+    FTL_PID_FILE="$( (grep "^PIDFILE=" $FTLCONFFILE || echo "$DEFAULT_PID_FILE") | cut -d"=" -f2-)"
+  else
+    # if there is no pihole-FTL.conf, use the default path
+    FTL_PID_FILE="$DEFAULT_PID_FILE"
+  fi
+
+  if [ -s "${FTL_PID_FILE}" ]; then
+    # -s: FILE exists and has a size greater than zero
+    FTL_PID="$(cat "$FTL_PID_FILE")"
+    # Exploit prevention: unset the variable if there is malicious content
+    # Verify that the value read from the file is numeric
+    expr "$FTL_PID" : "[^[:digit:]]" > /dev/null && unset FTL_PID
+  fi
+
+  # If FTL is not running, or the PID file contains malicious stuff, substitute
+  # negative PID to signal this to the caller
+  echo "${FTL_PID:=-1}"
+}
+
+
 is_running() {
-  pgrep -xo "pihole-FTL" > /dev/null
+  FTL_PID="$(getFTLPID)"
+  if [ "$FTL_PID" -eq "-1" ]; then
+    return 1
+  else
+    echo 0
+  fi
```

Do not trust that the process is running when a valid PID file exists and the other way round. Instead check for the `/proc/<pid>` directory. Obtain the PID only once and reuse the variable to reduce overhead on service stop. Also (silently) return the result instead of printing "0":
```suggestion
  if [ -z "$FTL_PID" ]; then
    FTL_PID="$(getFTLPID)"
  fi
  if [ -d "/proc/$FTL_PID" ]; then
    return 0
  fi
  return 1
```